api/admin/controller.py

```diff
@@ -1030,6 +1046,228 @@ def edit_classifications(self, identifier_type, identifier):
+
+        if image_url and not image_file:
+            image_file = StringIO(urllib.urlopen(image_url).read())
+        
+        image = Image.open(image_file)
+        result = self._validate_cover_image(image)
+        if isinstance(result, ProblemDetail):
+            return result
+
+        cover_href = None
+
+        if title_position != "none":
+            original_href = image_url
+            original_buffer = StringIO()
+            image.save(original_buffer, format="PNG")
+            original_content = original_buffer.getvalue()
+            if not original_href:
+                original_href = Hyperlink.generic_uri(data_source, work.presentation_edition.primary_identifier, Hyperlink.IMAGE, content=original_content)
+                
+            self._process_cover_image(work, image, title_position)
+
+            original = LinkData(
+                Hyperlink.IMAGE, original_href, rights_uri=rights_uri,
+                rights_explanation=rights_explanation, content=original_content,
+            )
+            derivation_settings = dict(title_position=title_position)
+        else:
+            original = None
+            derivation_settings = None
+            cover_href = image_url
+
+        buffer = StringIO()
+        image.save(buffer, format="PNG")
+        content = buffer.getvalue()
+
+        if not cover_href:
+            cover_href = Hyperlink.generic_uri(data_source, work.presentation_edition.primary_identifier, Hyperlink.IMAGE, content=content)
+
+        cover_data = LinkData(
+            Hyperlink.IMAGE, href=cover_href,
+            media_type=Representation.PNG_MEDIA_TYPE,
+            content=content, rights_uri=rights_uri,
+            rights_explanation=rights_explanation,
+            original=original, derivation_settings=derivation_settings,
+        )
+
+        presentation_policy = PresentationCalculationPolicy(
+            choose_edition=False,
+            set_edition_metadata=False,
+            classify=False,
+            choose_summary=False,
+            calculate_quality=False,
+            choose_cover=True,
+            regenerate_opds_entries=True,
+            update_search_index=False,
+        )
+
+        replacement_policy = ReplacementPolicy(
+            links=True,
+            # link_content is false because we already have the content.
+            # We don't want the metadata layer to try to fetch it again.
+            link_content=False,
+            mirror=mirror,
+            presentation_calculation_policy=presentation_policy,
+        )
+
+        metadata = Metadata(data_source, links=[cover_data])
+        metadata.apply(work.presentation_edition,
+                       collection,
+                       replace=replacement_policy)
+
+        # metadata.apply only updates the edition, so we also need
+        # to update the work.
+        work.calculate_presentation(policy=presentation_policy)
```

Comment:
There may be some extra work happening in metadata.apply since I reused the policy.